portfolio/src/main/java/com/google/sps/data/CalendarEventsGroup.java

```diff
@@ -0,0 +1,103 @@
+   *     user.
+   * @param startTime: start of possible scheduling blocks, of type Instant. This is the earlier of
+   *     working hour start time and the current timestamp.
+   * @param endTime: end of possible scheduling blocks, of type Instant.
+   */
+  public CalendarEventsGroup(
+      List<CalendarEvent> events, Instant overallStartTime, Instant overallEndTime) {
+    if (events == null) {
+      throw new IllegalArgumentException("Events cannot be null");
+    }
+    if (overallStartTime == null) {
+      throw new IllegalArgumentException("Overall start of scheduling hours cannot be null");
+    }
+    if (overallEndTime == null) {
+      throw new IllegalArgumentException("Overall end of scheduling hours cannot be null");
+    }
+    if (overallStartTime.isAfter(overallEndTime)) {
+      throw new IllegalArgumentException("Start time cannot be after end time");
+    }
+    this.events = Lists.newArrayList(events);
+    this.overallStartTime = overallStartTime;
+    this.overallEndTime = overallEndTime;
+  }
+
+  /**
+   * Calculates a TimeRange list which represents the periods of time that are empty of events and
+   * lie completely inside the possible scheduling hours. Originally written by tomasalvarez,
+   * modified by hollyyuqizheng.
+   */
+  public List<TimeRange> getFreeTimeRanges() {
+    Collections.sort(events, sortByEventStartTimeAscending);
+
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // This represents the earliest time that we can schedule a window for the
+    // meeting. As events are processed, this changes to their end times.
+    Instant earliestNonScheduledInstant = overallStartTime;
+
+    for (CalendarEvent event : events) {
+      // Make sure that there is some time between the events and it is not
+      // later than the person's scheduling hours' ending time.
+      if (event.getStartTime().isAfter(earliestNonScheduledInstant)) {
+        if (event.getStartTime().isAfter(overallEndTime)) {
+          possibleTimes.add(TimeRange.fromStartEnd(earliestNonScheduledInstant, overallEndTime));
+        } else {
+          possibleTimes.add(
+              TimeRange.fromStartEnd(earliestNonScheduledInstant, event.getStartTime()));
+        }
+      }
+      // Check if the earliest non scheduled time needs to be shifted to later.
+      if (earliestNonScheduledInstant.isBefore(event.getEndTime())) {
+        earliestNonScheduledInstant = event.getEndTime();
+      }
+    }
+    // The end of the work hours is potentially never included so we check.
+    if (overallEndTime.isAfter(earliestNonScheduledInstant)) {
+      possibleTimes.add(TimeRange.fromStartEnd(earliestNonScheduledInstant, overallEndTime));
+    }
+
+    return possibleTimes;
+  }
+
+  /**
+   * Adds a new event into the event group. Returns the new collection of events, sorted based on
+   * start time.
+   */
+  public List<CalendarEvent> addNewEvent(CalendarEvent event) {
```

Comment:
Consider making it return void. I think the caller probably only needs the free time ranges as output as far as this class is concerned.

Reply:
Updated.